src/main/java/org/folio/processing/mapping/mapper/writer/marc/MarcRecordModifier.java

```diff
@@ -889,18 +890,52 @@ private boolean isNotProtected(ControlField field) {
-        setting -> setting.getSubfield().equals(ANY_STRING) || field.getSubfield(setting.getSubfield().charAt(0)) != null)
-      .noneMatch(setting -> setting.getData().equals(ANY_STRING) || setting.getData()
-        .equals(field.getSubfield(setting.getSubfield().charAt(0)).getData()));
+      .filter(setting -> matchesField(setting, field))
+      .filter(setting -> matchesIndicator1(setting, field))
+      .filter(setting -> matchesIndicator2(setting, field))
+      .filter(setting -> hasSubfield(setting, field))
+      .noneMatch(setting -> matchesData(setting, field));
+  }
+
+  private boolean matchesField(MarcFieldProtectionSetting setting, DataField field) {
+    LOGGER.trace("matchesField:: field={}, setting: field={}", field.getTag(), setting.getField());
+    return setting.getField().equals(ANY_STRING) || setting.getField().equals(field.getTag());
+  }
+
+  private boolean matchesIndicator1(MarcFieldProtectionSetting setting, DataField field) {
+    LOGGER.trace("matchesIndicator1:: field={} | setting: indicator1={}", field.getTag(), setting.getIndicator1());
+    return setting.getIndicator1().equals(ANY_STRING) ||
+      (isNotEmpty(setting.getIndicator1()) ? setting.getIndicator1().charAt(0) : BLANK_SUBFIELD_CODE) == field.getIndicator1();
+  }
+
+  private boolean matchesIndicator2(MarcFieldProtectionSetting setting, DataField field) {
+    LOGGER.trace("matchesIndicator2:: field={} | setting: indicator2={}", field.getTag(), setting.getIndicator2());
+    return setting.getIndicator2().equals(ANY_STRING) ||
+      (isNotEmpty(setting.getIndicator2()) ? setting.getIndicator2().charAt(0) : BLANK_SUBFIELD_CODE) == field.getIndicator2();
+  }
+
+  private boolean hasSubfield(MarcFieldProtectionSetting setting, DataField field) {
+    LOGGER.trace("hasSubfield:: field={} | setting: subfield={}", field.getTag(), setting.getSubfield());
+    return setting.getSubfield().equals(ANY_STRING) || field.getSubfield(setting.getSubfield().charAt(0)) != null;
+  }
+
+  private boolean matchesData(MarcFieldProtectionSetting setting, DataField field) {
+    LOGGER.trace("matchesData:: field={} | setting: subfield={}, data={}", field.getTag(), setting.getSubfield(), setting.getData());
+    if (setting.getSubfield().charAt(0) == ANY_CHAR) {
+      return setting.getData().equals(ANY_STRING) ||
+        field.getSubfields().stream().anyMatch(subfield -> dataMatches(setting, subfield));
+    } else {
+      return Optional.ofNullable(field.getSubfield(setting.getSubfield().charAt(0)))
+        .map(subfield -> dataMatches(setting, subfield))
+        .orElse(false);
+    }
+  }
+
+  private boolean dataMatches(MarcFieldProtectionSetting setting, Subfield subfield) {
+    LOGGER.trace("dataMatches:: subfield: code={}, data={} | setting: subfield={}, data={}", subfield.getCode(), subfield.getData(), setting.getSubfield(), setting.getData());
+    return setting.getData().equals(ANY_STRING) || setting.getData().equals(subfield.getData());
```

Comment:
```suggestion
    return setting.getData().equals(subfield.getData());
```
propose to remove this `setting.getData().equals(ANY_STRING)` check here because we do it earlier in the matchesData() method

Reply:
just found that here this check can be used when setting.subfield != '*' in that case this check can be executed. Thank you @Aliaksandr-Fedasiuk!

Reply:
Thanks @RuslanLavrov, fixed